packages/compiler/core/parser.ts

```diff
@@ -849,11 +861,11 @@ export function parse(code: string | SourceFile, options: ParseOptions = {}): Ca
       options.push(expr);
     }
 
-    return {
+    return createNode({
       kind: SyntaxKind.UnionExpression,
       options,
       ...finishNode(pos),
-    };
+    }) as UnionExpressionNode;
```

Comment:
Why did all these casts become necessary?

Reply:
I'm not exactly sure, I was going to poke at that some today. For some reason TS started inferring the return type for the T of createNode (expression in this case, i.e. type information is flowing into createNode) as opposed to the more specific type that could be inferred from the provided kind.

Reply:
No error with that form, it's effectively the same as the cast though I think. I tried to find some type system magic that would fix this issue, e.g. by making the generic parameter `T extends Node["kind"]`, but I couldn't get it to work well.